lib/ruby_lsp/document.rb

```diff
@@ -190,10 +198,31 @@ def locate(node, char_position, node_types: [])
       # though the class/module node does indeed enclose the target, because it would lead to incorrect behavior
       if closest.is_a?(Prism::ConstantReadNode) || closest.is_a?(Prism::ConstantPathNode)
         last_level = nesting.last
-        nesting.pop if last_level && last_level.constant_path == closest
+
+        if (last_level.is_a?(Prism::ModuleNode) || last_level.is_a?(Prism::ClassNode)) &&
+            last_level.constant_path == closest
+          nesting.pop
+        end
+      end
+
+      complete_nesting = []
+      surrounding_method = T.let(nil, T.nilable(String))
+
+      nesting.each do |node|
+        case node
+        when Prism::ClassNode, Prism::ModuleNode
+          complete_nesting << node.constant_path.slice
+        when Prism::SingletonClassNode
+          complete_nesting << "<Class:#{complete_nesting.last}>"
+        when Prism::DefNode
+          surrounding_method = node.name.to_s
+          next unless node.receiver.is_a?(Prism::SelfNode)
```

Comment:
nit: this `next` can be called first.

Reply:
It can't because we want to remember the surrounding instance methods as well. That will be necessary to provide hover, signature help and go to definition on `super` calls.

```ruby
class Foo < Bar
  def something
    # When we find this `super` call, we need to know that we're inside
    # `something` because that's what we will be searching for in the ancestors
    # of `Foo`
    super
  end
end
```

Reply:
Oh sorry I misread the scope of the local.